File: sdk/python/alp_sdk/p2p.py
"""ALP Decentralized Coordination (v18.1.0 — V14 The Sovereign Era).

P2P swarm coordination without a central coordinator:

* ``P2PNode``         — a peer in the decentralized swarm.
* ``P2PSwarm``        — gossip-based state sync, direct negotiation, ad-hoc federations.
* ``GossipProtocol``  — best-effort rumor spreading for state synchronization.
* ``DHT``             — lightweight distributed hash table for agent discovery.

Mirrors the planned ``parser/src/p2p.ts`` surface; tests live in
``sdk/python/tests/test_p2p.py``.
"""

import hashlib
import json
import os
import time
import uuid
from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Callable, Dict, List, Optional, Set


class AgentStatus(str, Enum):
    ACTIVE = "active"
    STANDBY = "standby"
    FAILED = "failed"
    REPLACED = "replaced"


P2P_DIR = ".p2p"
PEERS_FILE = "peers.jsonl"
STATE_FILE = "swarm_state.json"


def p2p_dir(alp_dir: str) -> str:
    return os.path.join(alp_dir, P2P_DIR)


def peers_path(alp_dir: str) -> str:
    return os.path.join(p2p_dir(alp_dir), PEERS_FILE)


def swarm_state_path(alp_dir: str) -> str:
    return os.path.join(p2p_dir(alp_dir), STATE_FILE)


@dataclass
class P2PNode:
    node_id: str
    agent_id: str
    capabilities: List[str] = field(default_factory=list)
    address: str = ""
    last_seen: str = ""
    metadata: Dict[str, Any] = field(default_factory=dict)

    def __post_init__(self):
        if not self.last_seen:
            self.last_seen = _now_iso()

    def to_dict(self) -> Dict[str, Any]:
        return {
            "node_id": self.node_id,
            "agent_id": self.agent_id,
            "capabilities": self.capabilities,
            "address": self.address,
            "last_seen": self.last_seen,
            "metadata": self.metadata,
        }

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "P2PNode":
        return cls(
            node_id=d["node_id"],
            agent_id=d["agent_id"],
            capabilities=d.get("capabilities", []),
            address=d.get("address", ""),
            last_seen=d.get("last_seen", ""),
            metadata=d.get("metadata", {}),
        )


@dataclass
class TaskAssignment:
    task_id: str
    agent_id: str
    workflow_id: str
    status: str = "assigned"
    retries: int = 0
    assigned_at: str = ""

    def __post_init__(self):
        if not self.assigned_at:
            self.assigned_at = _now_iso()

    def to_dict(self) -> Dict[str, Any]:
        return {
            "task_id": self.task_id,
            "agent_id": self.agent_id,
            "workflow_id": self.workflow_id,
            "status": self.status,
            "retries": self.retries,
            "assigned_at": self.assigned_at,
        }


@dataclass
class GossipMessage:
    topic: str
    payload: Dict[str, Any]
    source_node: str
    timestamp: str = ""
    ttl: int = 3

    def __post_init__(self):
        if not self.timestamp:
            self.timestamp = _now_iso()

    def to_dict(self) -> Dict[str, Any]:
        return {
            "topic": self.topic,
            "payload": self.payload,
            "source_node": self.source_node,
            "timestamp": self.timestamp,
            "ttl": self.ttl,
        }


class DHT:
    """Lightweight distributed hash table for agent discovery."""

    def __init__(self):
        self._table: Dict[str, P2PNode] = {}

    def register(self, node: P2PNode) -> None:
        key = self._node_key(node.agent_id)
        self._table[key] = node

    def resolve(self, agent_id: str) -> Optional[P2PNode]:
        return self._table.get(self._node_key(agent_id))

    def remove(self, agent_id: str) -> None:
        self._table.pop(self._node_key(agent_id), None)

    def find_by_capability(self, capability: str) -> List[P2PNode]:
        return [n for n in self._table.values() if capability in n.capabilities]

    def all_nodes(self) -> List[P2PNode]:
        return list(self._table.values())

    @staticmethod
    def _node_key(agent_id: str) -> str:
        return hashlib.sha256(agent_id.encode()).hexdigest()[:16]


class GossipProtocol:
    """Best-effort rumor spreading for swarm state synchronization."""

    def __init__(self, fanout: int = 3):
        self.fanout = fanout
        self._seen: Set[str] = set()

    def spread(self, message: GossipMessage, peers: List[P2PNode]) -> List[GossipMessage]:
        forwarded: List[GossipMessage] = []
        msg_id = self._message_id(message)
        if msg_id in self._seen:
            return forwarded
        self._seen.add(msg_id)
        if message.ttl <= 0:
            return forwarded
        for peer in peers[: self.fanout]:
            next_payload = dict(message.payload)
            next_payload["_forwarded_to"] = peer.node_id
            forwarded.append(
                GossipMessage(
                    topic=message.topic,
                    payload=next_payload,
                    source_node=message.source_node,
                    timestamp=message.timestamp,
                    ttl=message.ttl - 1,
                )
            )
        return forwarded

    def _message_id(self, message: GossipMessage) -> str:
        raw = json.dumps(message.to_dict(), sort_keys=True, default=str).encode()
        return hashlib.sha256(raw).hexdigest()


class P2PSwarm:
    """Decentralized swarm coordination without a central coordinator.

    Usage::

        swarm = P2PSwarm(alp_dir="/path/to/.alp")
        swarm.join(P2PNode("n1", "agent-1", capabilities=["build"]))
        swarm.join(P2PNode("n2", "agent-2", capabilities=["test"]))
        swarm.gossip(GossipMessage("task.assign", {"task_id": "t1", "agent": "agent-1"}, "n1"))
        report = swarm.run(executor=lambda task_id, agent_id: None, tasks=[{"task_id": "t1"}])
    """

    def __init__(
        self,
        alp_dir: str,
        fanout: int = 3,
        heartbeat_timeout: float = 30.0,
        max_retries: int = 2,
    ):
        self.alp_dir = alp_dir
        self.fanout = fanout
        self.heartbeat_timeout = heartbeat_timeout
        self.max_retries = max_retries
        self.dht = DHT()
        self.gossip_protocol = GossipProtocol(fanout=fanout)
        self._assignments: List[TaskAssignment] = []
        self._reports: Dict[str, P2PReport] = {}
        self._messages: List[GossipMessage] = []

    def join(self, node: P2PNode) -> None:
        self.dht.register(node)
        self._persist_peer(node)

    def leave(self, agent_id: str) -> None:
        self.dht.remove(agent_id)
        self._remove_peer(agent_id)

    def gossip(self, message: GossipMessage) -> List[GossipMessage]:
        peers = self.dht.all_nodes()
        forwarded = self.gossip_protocol.spread(message, peers)
        self._messages.append(message)
        self._messages.extend(forwarded)
        return forwarded

    def discover(self, capability: str) -> List[P2PNode]:
        return self.dht.find_by_capability(capability)

    def assign_task(self, task_id: str, agent_id: str, workflow_id: str = "_default") -> Optional[TaskAssignment]:
        node = self.dht.resolve(agent_id)
        if not node:
            return None
        assignment = TaskAssignment(task_id=task_id, agent_id=agent_id, workflow_id=workflow_id)
        self._assignments.append(assignment)
        return assignment

    def _execute_task(self, executor: Callable[[str, str], Any], task_id: str, agent_id: str) -> bool:
        attempt = 0
        while attempt <= self.max_retries:
            try:
                executor(task_id, agent_id)
                return True
            except Exception:
                attempt += 1
                if attempt > self.max_retries:
                    return False
        return False

    def run(self, executor: Callable[[str, str], Any], tasks: List[Dict[str, Any]], workflow_id: str = "_default") -> "P2PReport":
        report = P2PReport(swarm_id=workflow_id)
        self._reports[workflow_id] = report

        for task in tasks:
            task_id = task.get("task_id", "")
            assignment = next((a for a in self._assignments if a.task_id == task_id), None)
            if not assignment:
                continue
            agent_id = assignment.agent_id
            node = self.dht.resolve(agent_id)
            if not node:
                report.add_action({"type": "task_failed", "task_id": task_id, "reason": "node not found"})
                continue

            success = self._execute_task(executor, task_id, agent_id)
            if success:
                assignment.status = "completed"
                report.add_action({"type": "task_completed", "task_id": task_id, "agent_id": agent_id})
            else:
                node.status = AgentStatus.FAILED
                assignment.status = "failed"
                report.add_action({"type": "task_failed", "task_id": task_id, "agent_id": agent_id})

        return report

    def get_report(self, workflow_id: Optional[str] = None) -> Optional["P2PReport"]:
        key = workflow_id or self.alp_dir
        return self._reports.get(key)

    def _persist_peer(self, node: P2PNode) -> None:
        try:
            d = p2p_dir(self.alp_dir)
            if not os.path.exists(d):
                os.makedirs(d, exist_ok=True)
            with open(peers_path(self.alp_dir), "a", encoding="utf-8") as f:
                f.write(json.dumps(node.to_dict()) + "\n")
        except Exception:
            pass

    def _remove_peer(self, node_id: str) -> None:
        try:
            p = peers_path(self.alp_dir)
            if not os.path.exists(p):
                return
            kept = []
            with open(p, "r", encoding="utf-8") as f:
                for line in f:
                    try:
                        entry = json.loads(line.strip())
                    except json.JSONDecodeError:
                        kept.append(line)
                        continue
                    if entry.get("node_id") != node_id:
                        kept.append(json.dumps(entry) + "\n")
            with open(p, "w", encoding="utf-8") as f:
                f.writelines(kept)
        except Exception:
            pass


@dataclass
class P2PReport:
    swarm_id: str
    actions: List[Dict[str, Any]] = field(default_factory=list)
    started_at: str = ""
    finished_at: str = ""

    def __post_init__(self):
        if not self.started_at:
            self.started_at = _now_iso()

    def add_action(self, action: Dict[str, Any]) -> None:
        self.actions.append(action)

    def to_dict(self) -> Dict[str, Any]:
        return {
            "swarm_id": self.swarm_id,
            "started_at": self.started_at,
            "finished_at": self.finished_at or _now_iso(),
            "total_actions": len(self.actions),
            "completed": sum(1 for a in self.actions if a.get("type") == "task_completed"),
            "failed": sum(1 for a in self.actions if a.get("type") == "task_failed"),
            "actions": self.actions,
        }


def _now_iso() -> str:
    from datetime import datetime, timezone
    return datetime.now(timezone.utc).isoformat()

File: parser/src/p2p.ts
/** ALP Decentralized Coordination (v18.1.0 — V14 The Sovereign Era).
 *
 * P2P swarm coordination without a central coordinator:
 *
 * - `P2PNode`         — a peer in the decentralized swarm.
 * - `P2PSwarm`        — gossip-based state sync, direct negotiation, ad-hoc federations.
 * - `GossipProtocol`  — best-effort rumor spreading for state synchronization.
 * - `DHT`             — lightweight distributed hash table for agent discovery.
 *
 * Mirrors `sdk/python/alp_sdk/p2p.py`.
 */

import * as crypto from 'node:crypto'
import * as fs from 'fs'
import * as path from 'path'

export interface P2PNodeData {
  node_id: string
  agent_id: string
  capabilities: string[]
  address: string
  last_seen: string
  metadata: Record<string, any>
}

export class P2PNode {
  node_id: string
  agent_id: string
  capabilities: string[]
  address: string
  last_seen: string
  metadata: Record<string, any>

  constructor(node_id: string, agent_id: string, capabilities: string[] = [], address = '', last_seen = '', metadata: Record<string, any> = {}) {
    this.node_id = node_id
    this.agent_id = agent_id
    this.capabilities = capabilities
    this.address = address
    this.last_seen = last_seen || new Date().toISOString()
    this.metadata = metadata
  }

  toDict(): P2PNodeData {
    return {
      node_id: this.node_id,
      agent_id: this.agent_id,
      capabilities: this.capabilities,
      address: this.address,
      last_seen: this.last_seen,
      metadata: this.metadata,
    }
  }

  static fromDict(d: Record<string, any>): P2PNode {
    return new P2PNode(
      d.node_id ?? d['node_id'],
      d.agent_id ?? d['agent_id'],
      d.capabilities ?? [],
      d.address ?? '',
      d.last_seen ?? '',
      d.metadata ?? {},
    )
  }
}

export interface TaskAssignmentData {
  task_id: string
  agent_id: string
  workflow_id: string
  status: string
  retries: number
  assigned_at: string
}

export class TaskAssignment {
  task_id: string
  agent_id: string
  workflow_id: string
  status: string
  retries: number
  assigned_at: string

  constructor(task_id: string, agent_id: string, workflow_id: string, status = 'assigned', retries = 0, assigned_at = '') {
    this.task_id = task_id
    this.agent_id = agent_id
    this.workflow_id = workflow_id
    this.status = status
    this.retries = retries
    this.assigned_at = assigned_at || new Date().toISOString()
  }

  toDict(): TaskAssignmentData {
    return {
      task_id: this.task_id,
      agent_id: this.agent_id,
      workflow_id: this.workflow_id,
      status: this.status,
      retries: this.retries,
      assigned_at: this.assigned_at,
    }
  }
}

export interface GossipMessageData {
  topic: string
  payload: Record<string, any>
  source_node: string
  timestamp: string
  ttl: number
}

export class GossipMessage {
  topic: string
  payload: Record<string, any>
  source_node: string
  timestamp: string
  ttl: number

  constructor(topic: string, payload: Record<string, any>, source_node: string, timestamp = '', ttl = 3) {
    this.topic = topic
    this.payload = payload
    this.source_node = source_node
    this.timestamp = timestamp || new Date().toISOString()
    this.ttl = ttl
  }

  toDict(): GossipMessageData {
    return {
      topic: this.topic,
      payload: this.payload,
      source_node: this.source_node,
      timestamp: this.timestamp,
      ttl: this.ttl,
    }
  }
}

export class DHT {
  private table: Map<string, P2PNode> = new Map()

  private nodeKey(agent_id: string): string {
    return crypto.createHash('sha256').update(agent_id).digest('hex').slice(0, 16)
  }

  register(node: P2PNode): void {
    this.table.set(this.nodeKey(node.agent_id), node)
  }

  resolve(agent_id: string): P2PNode | undefined {
    return this.table.get(this.nodeKey(agent_id))
  }

  remove(agent_id: string): void {
    this.table.delete(this.nodeKey(agent_id))
  }

  findByCapability(capability: string): P2PNode[] {
    return Array.from(this.table.values()).filter(n => n.capabilities.includes(capability))
  }

  allNodes(): P2PNode[] {
    return Array.from(this.table.values())
  }
}

export class GossipProtocol {
  private fanout: number
  private seen: Set<string> = new Set()

  constructor(fanout = 3) {
    this.fanout = fanout
  }

  spread(message: GossipMessage, peers: P2PNode[]): GossipMessage[] {
    const forwarded: GossipMessage[] = []
    const msg_id = this.messageId(message)
    if (this.seen.has(msg_id)) return forwarded
    this.seen.add(msg_id)
    if (message.ttl <= 0) return forwarded

    const fanoutPeers = peers.slice(0, this.fanout)
    for (const peer of fanoutPeers) {
      const next_payload = { ...message.payload, _forwarded_to: peer.node_id }
      forwarded.push(new GossipMessage(message.topic, next_payload, message.source_node, message.timestamp, message.ttl - 1))
    }
    return forwarded
  }

  private messageId(message: GossipMessage): string {
    const raw = JSON.stringify(message.toDict())
    return crypto.createHash('sha256').update(raw).digest('hex')
  }
}

export interface P2PReportData {
  swarm_id: string
  actions: Array<Record<string, any>>
  started_at: string
  finished_at: string
  total_actions: number
  completed: number
  failed: number
}

export class P2PReport {
  swarm_id: string
  actions: Array<Record<string, any>> = []
  started_at: string
  finished_at: string = ''

  constructor(swarm_id: string) {
    this.swarm_id = swarm_id
    this.started_at = new Date().toISOString()
  }

  addAction(action: Record<string, any>): void {
    this.actions.push(action)
  }

  toDict(): P2PReportData {
    return {
      swarm_id: this.swarm_id,
      actions: this.actions,
      started_at: this.started_at,
      finished_at: this.finished_at || new Date().toISOString(),
      total_actions: this.actions.length,
      completed: this.actions.filter(a => a.type === 'task_completed').length,
      failed: this.actions.filter(a => a.type === 'task_failed').length,
    }
  }
}

export class P2PSwarm {
  private alp_dir: string
  private fanout: number
  private heartbeat_timeout: number
  private max_retries: number
  private dht: DHT
  private gossip_protocol: GossipProtocol
  private assignments: TaskAssignment[] = []
  private reports: Map<string, P2PReport> = new Map()
  private messages: GossipMessage[] = []

  constructor(
    alp_dir: string,
    fanout = 3,
    heartbeat_timeout = 30_000,
    max_retries = 2,
  ) {
    this.alp_dir = alp_dir
    this.fanout = fanout
    this.heartbeat_timeout = heartbeat_timeout
    this.max_retries = max_retries
    this.dht = new DHT()
    this.gossip_protocol = new GossipProtocol(fanout)
  }

  join(node: P2PNode): void {
    this.dht.register(node)
    this.persist_peer(node)
  }

  leave(agent_id: string): void {
    this.dht.remove(agent_id)
    this.remove_peer(agent_id)
  }

  gossip(message: GossipMessage): GossipMessage[] {
    const peers = this.dht.allNodes()
    const forwarded = this.gossip_protocol.spread(message, peers)
    this.messages.push(message)
    this.messages.push(...forwarded)
    return forwarded
  }

  discover(capability: string): P2PNode[] {
    return this.dht.findByCapability(capability)
  }

  assign_task(task_id: string, agent_id: string, workflow_id = '_default'): TaskAssignment | undefined {
    const node = this.dht.resolve(agent_id)
    if (!node) return undefined
    const assignment = new TaskAssignment(task_id, agent_id, workflow_id)
    this.assignments.push(assignment)
    return assignment
  }

  run(executor: (task_id: string, agent_id: string) => void, tasks: Array<Record<string, any>>, workflow_id = '_default'): P2PReport {
    const report = new P2PReport(workflow_id)
    this.reports.set(workflow_id, report)

    for (const task of tasks) {
      const task_id = task.task_id ?? task['task_id'] ?? ''
      const assignment = this.assignments.find(a => a.task_id === task_id)
      if (!assignment) continue

      const node = this.dht.resolve(assignment.agent_id)
      if (!node) {
        report.addAction({ type: 'task_failed', task_id, reason: 'node not found' })
        continue
      }

      let attempt = 0
      let success = false
      while (attempt <= this.max_retries) {
        try {
          executor(task_id, assignment.agent_id)
          success = true
          break
        } catch {
          attempt++
        }
      }

      if (success) {
        assignment.status = 'completed'
        report.addAction({ type: 'task_completed', task_id, agent_id: assignment.agent_id })
      } else {
        assignment.status = 'failed'
        report.addAction({ type: 'task_failed', task_id, agent_id: assignment.agent_id })
      }
    }

    return report
  }

  getReport(workflow_id?: string): P2PReport | undefined {
    const key = workflow_id ?? this.alp_dir
    return this.reports.get(key)
  }

  private persist_peer(node: P2PNode): void {
    try {
      const d = path.join(this.alp_dir, '.p2p')
      if (!fs.existsSync(d)) {
        fs.mkdirSync(d, { recursive: true })
      }
      fs.appendFileSync(path.join(d, 'peers.jsonl'), JSON.stringify(node.toDict()) + '\n')
    } catch {
      // best-effort
    }
  }

  private remove_peer(node_id: string): void {
    try {
      const p = path.join(this.alp_dir, '.p2p', 'peers.jsonl')
      if (!fs.existsSync(p)) return
      const lines = fs.readFileSync(p, 'utf-8').split('\n').filter(Boolean)
      const kept: string[] = []
      for (const line of lines) {
        try {
          const entry = JSON.parse(line)
          if (entry.node_id !== node_id) {
            kept.push(line)
          }
        } catch {
          kept.push(line)
        }
      }
      fs.writeFileSync(p, kept.join('\n') + (kept.length ? '\n' : ''))
    } catch {
      // best-effort
    }
  }
}
